### ISSUE TITLE
awful.keygrabber() keybindings do not work with keycodes

### ISSUE THREAD

**Issue by HactarCE** [awesomeWM/awesome#2807]
Output of `awesome --version`:

```
awesome v4.3 (Too long)
 • Compiled against Lua 5.3.5 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.6
 • LGI version: 0.9.2
```

**How to reproduce the issue:**

Example 1:
```lua
awful.keygrabber{keybindings = {{{}, 'w', function() print('w pressed') end}}}:start()
```

Example 2:
```lua
awful.keygrabber{keybindings = {{{}, '#25', function() print('w pressed') end}}}:start()
```

**Actual result:**

Only example #1 prints `w pressed` when pressing the `w` key with the Qwerty layout.

**Expected result:**

Both examples should print `w pressed` when pressing the `w` key in the Qwerty layout.

**Comment by HactarCE** [comment]
Is it possible to resolve this within Awesome's Lua library, or does it have to happen at the C level? If it can be done in Lua I'd be happy to give it a shot.

**Comment by Elv13** [comment]
> Is it possible to resolve this within Awesome's Lua library

Probably. If I remember, `fake_input` allows keycodes, but I am not sure this is documented or working. In the end, the keygrabber code just emulates the normal keybindings API, it doesn't share much with the "original", hence your issue.

**Comment by psychon** [comment]
> Is it possible to resolve this within Awesome's Lua library, or does it have to happen at the C level?

TL;DR: Required changes to the C code.

The C-code calls the keygrabber callback function with the key that was pressed, modifiers, and state (press/release). At this point, the keycode was already translated to a keysym. So, this requires changes to the C code. Untested quick hack that adds a new argument to the callback function with the keycode:
```diff
diff --git a/event.c b/event.c
index 989c0110c..595cebc10 100644
--- a/event.c
+++ b/event.c
@@ -761,7 +761,7 @@ event_handle_key(xcb_key_press_event_t *ev)
         {
             lua_rawgeti(L, LUA_REGISTRYINDEX, globalconf.keygrabber);
 
-            if(!luaA_dofunction(L, 3, 0))
+            if(!luaA_dofunction(L, 4, 0))
             {
                 warn("Stopping keygrabber.");
                 luaA_keygrabber_stop(L);
diff --git a/keygrabber.c b/keygrabber.c
index d0c927fcd..1182213bc 100644
--- a/keygrabber.c
+++ b/keygrabber.c
@@ -106,6 +106,7 @@ keygrabber_handlekpress(lua_State *L, xcb_key_press_event_t *e)
         lua_pushliteral(L, "release");
         break;
     }
+    lua_pushinteger(L, ev->detail);
     return true;
 }
 
```

**Comment by HactarCE** [comment]
@psychon I cloned the latest from `master` and applied your patch (changing your `ev->detail` to `e->detail`) but it didn't seem to make a difference.

> In the end, the keygrabber code just emulates the normal keybindings API

I know I'm not particularly well-versed with Awesome's internals, and I understand there's quite a bit of legacy going on, but would it make sense to have all keybindings ultimately built on keygrabbers (with each keygrabber given an option to let through any keys it doesn't match)? Then everything relating to keybinds would be in one place. Just a thought.

**Comment by Elv13** [comment]
> but would it make sense to have all keybindings ultimately built on keygrabbers

No. The normal keybindings are implemented server side in a different way. The keygrabber forces extra round trips and require to pause them to send inputs to the clients. This would be very slow and there is an opportunity for race conditions before Awesome restarts global keygrabbing. Most games would break.

You might be interested in https://github.com/awesomeWM/awesome/pull/2508 , it makes some baby steps to turn `awful.key` and `awful.buttons` into real objects instead of tuples of `capi.key` and `capi.buttons`. Down the road this would be used to unify the user facing API. However the internals would continue to work as they currently do.

> but it didn't seem to make a difference

The patch above adds an extra parameters to the keygrabber function callback. It doesn't implements the new bits in `awful.keygrabber`.

**Comment by psychon** [comment]
> but it didn't seem to make a difference.

The patch only changes the C code, not the Lua code. The Lua function gets called with an extra argument, but the current Lua function just ignores this argument. For `awful.keygrabber`, something like this makes this extra argument visible, but does not use it in any way (and I do not have time right now to dig deeper into `awful.keygrabber`):
```diff
diff --git a/lib/awful/keygrabber.lua b/lib/awful/keygrabber.lua
index 2d076c672..2da4db529 100644
--- a/lib/awful/keygrabber.lua
+++ b/lib/awful/keygrabber.lua
@@ -178,7 +178,8 @@ end
 
 --END hack
 
-local function grabber(mod, key, event)
+local function grabber(mod, key, event, detail)
+    print("keygrabber called with arguments", mod, key, event, detail)
     for _, keygrabber_function in ipairs(grabbers) do
         -- continue if the grabber explicitly returns false
         if keygrabber_function(mod, key, event) ~= false then
```
Sorry for not being more helpful.
